R/request_functions.R

```diff
@@ -266,3 +261,86 @@ get_gbif_points = function(key) {
   print(res[1,2])
   return(res)
 }
+
+get_random_powo = function(){
+  
+  powo_fam_results <- tibble(
+    family=NA_character_,
+    accepted=NA_character_
+  )
+  
+  # use name full name to search API  
+  full_url =  paste("http://plantsoftheworldonline.org/api/1/search?f=accepted_names%2%2Cfamily_ff")
+  
+  # encode
+  full_url = utils::URLencode(full_url)
+  
+  # get raw json data
+  raw_data <- readLines(full_url, warn = "F", encoding = "UTF-8")
+  
+  # organise
+  rd = fromJSON(raw_data)
+  
+  # make data frame
+  results = rd$results
+  
+  #rd$totalResults
+  
+  # get IPNI ID
+  #results = mutate(results, IPNI_ID=str_extract(url, "(?<=names\\:)[\\d\\-]+$"))
+  
+  # only include these fields - you don't want synonym of
+  fam_results = select(results, colnames(powo_fam_results))
+  fam_results = subset(fam_results, accepted == "TRUE")
+  fam_results = subset(fam_results, select = family)
+  
+  # random page
+  #rand_page = sample(rd$totalPages,1)
+  
+  rand_page = sample(19,1)
+  
+  # run through cursors for n random pages
+  for (n in 1:rand_page){
+    
+    cursor = rd$cursor
+    full_url =  paste0("http://plantsoftheworldonline.org/api/1/search?cursor=",cursor)
+    raw_data <- readLines(full_url, warn = "F", encoding = "UTF-8")
+    rd = fromJSON(raw_data)
+  }
+
+  results = rd$results
+  #results = mutate(results, IPNI_ID=str_extract(url, "(?<=names\\:)[\\d\\-]+$"))
+  fam_results = select(results, family)
+    
+  # get random species from list
+  rand_fam = fam_results[sample(nrow(fam_results),1), ]
+  
+  # now get random species from that family
+  
+  #http://www.plantsoftheworldonline.org/?f=accepted_names%2Cspecies_f&q=Opiliaceae
+  
+  full_url = paste0("http://plantsoftheworldonline.org/api/1/search?f=accepted_names%2Cspecies_f&q=", rand_fam)
+  raw_data <- readLines(full_url, warn = "F", encoding = "UTF-8")
+  rd = fromJSON(raw_data)
+  
+  #rand_page = sample(50,1)
+  
+  #for (n in 1:rand_page){
```

Comment:
A simpler way of running this might be to generate a random number from the total number of species in the family, and then run the query until the number of results matches or exceeds the random number:

```
full_url <- paste0(paste0("http://plantsoftheworldonline.org/api/1/search?f=accepted_names%2Cspecies_f&page.size=480&q=", rand_fam)
raw_data <- readLines(full_url, warn=FALSE, encoding="UTF-8")
results <- fromJSON(raw_data)

random_species_number <- sample(results$totalResults, 1)
returned_species <- nrow(results$results)
cursor <- results$cursor

while (returned_species < random_species_number) {
  url <- paste0(full_url, "&cursor=", cursor, collapse="")
  raw_data <- readLines(url, warn=FALSE, encoding="UTF-8")
  results <- fromJSON(raw_data)
  returned_species <- results + nrow(results$results)
  cursor <- results$cursor
}
# we're only keeping the latest results from POWO, so need to change the random species number to match
random_species_number <- mod(random_species_number, 480)

random_species <- results$results[random_species_number, ]$name
```

Reply:
this works really well, thanks! I've applied it to genus and species levels. Usually takes 2-5 seconds to get a result.